service/history/query.go

```diff
@@ -23,131 +23,125 @@
+		queryResult          *shared.WorkflowQueryResult
+		failure              error
 	}
 )
 
 func newQuery(queryInput *shared.WorkflowQuery) query {
 	return &queryImpl{
-		id:          uuid.New(),
-		queryInput:  queryInput,
-		queryResult: nil,
-		termCh:      make(chan struct{}),
-		state:       queryStateBuffered,
+		id:         uuid.New(),
+		queryInput: queryInput,
+		termCh:     make(chan struct{}),
 	}
 }
 
-func (q *queryImpl) getQueryInternalState() *queryInternalState {
-	q.RLock()
-	defer q.RUnlock()
-
-	return &queryInternalState{
-		id:          q.id,
-		queryInput:  q.queryInput,
-		queryResult: q.queryResult,
-		state:       q.state,
-	}
+func (q *queryImpl) getQueryID() string {
+	return q.id
 }
 
 func (q *queryImpl) getQueryTermCh() <-chan struct{} {
 	return q.termCh
 }
 
-func (q *queryImpl) completeQuery(
-	queryResult *shared.WorkflowQueryResult,
-) error {
-
-	q.Lock()
-	defer q.Unlock()
+func (q *queryImpl) getQueryInput() *shared.WorkflowQuery {
+	return q.queryInput
+}
 
-	if q.terminalState() {
-		return errAlreadyInTerminalState
-	}
-	if err := validateQueryResult(queryResult); err != nil {
-		return err
+func (q *queryImpl) getTerminationState() (*queryTerminationState, error) {
+	ts := q.terminationState.Load()
+	if ts == nil {
+		return nil, errQueryNotInTerminalState
 	}
-	q.state = queryStateCompleted
-	q.queryResult = queryResult
-	close(q.termCh)
-	return nil
+	return ts.(*queryTerminationState), nil
 }
 
-func (q *queryImpl) unblockQuery() error {
-	q.Lock()
-	defer q.Unlock()
-
-	if q.terminalState() {
+func (q *queryImpl) setTerminationState(terminationState *queryTerminationState) error {
+	if err := q.validateTerminationState(terminationState); err != nil {
+		return err
+	}
+	currTerminationState, _ := q.getTerminationState()
+	if currTerminationState != nil {
 		return errAlreadyInTerminalState
 	}
-	q.state = queryStateUnblocked
+	q.terminationState.Store(terminationState)
 	close(q.termCh)
 	return nil
 }
 
-func (q *queryImpl) terminalState() bool {
-	return q.state == queryStateCompleted || q.state == queryStateUnblocked
-}
-
-func validateQueryResult(
-	queryResult *shared.WorkflowQueryResult,
+func (q *queryImpl) validateTerminationState(
+	terminationState *queryTerminationState,
 ) error {
-
-	if queryResult == nil {
-		return errQueryResultIsNil
+	if terminationState == nil {
+		return errTerminationStateInvalid
 	}
-	validAnswered := queryResult.GetResultType().Equals(shared.QueryResultTypeAnswered) &&
-		queryResult.Answer != nil &&
-		queryResult.ErrorMessage == nil
-
-	validFailed := queryResult.GetResultType().Equals(shared.QueryResultTypeFailed) &&
-		queryResult.Answer == nil &&
-		queryResult.ErrorMessage != nil
-
-	if !validAnswered && !validFailed {
-		return errQueryResultIsInvalid
+	switch terminationState.queryTerminationType {
+	case queryTerminationTypeCompleted:
+		if terminationState.queryResult == nil || terminationState.failure != nil {
```

Comment:
This changes the error for nil queryResult from errQueryResultIsNil to errTerminationStateInvalid. Is it intentional?